data/transactions/logic/eval.go

```diff
@@ -2015,18 +2025,35 @@ func opB(cx *EvalContext) error {
 }
 
 func opCallSub(cx *EvalContext) error {
-	cx.callstack = append(cx.callstack, cx.pc+3)
-	return opB(cx)
+	cx.callstack = append(cx.callstack, frame{
+		retpc:  cx.pc + 3, // retpc is pc _after_ the callsub
+		height: len(cx.stack),
+	})
+	err := opB(cx)
+	if cx.nextpc < len(cx.program) && cx.program[cx.nextpc] == 0xf0 {
+		cx.fromCallsub = true
+	}
```

Comment:
I should write a comment, especially on the `0xf0`. The deal is that we only indicate that we're coming from a callSub if we know we're jumping to a `proto` (that's `0xf0`).  If we did not do this check, then we would set the flag, jump somewhere, and then, maybe 10 instructions later, invoke `proto` and it would be allowed because the flag is set.

This is really a little handshake between the `callsub` and `proto`. Done this way, we don't have to check and clear the `fromCallsub` flag in every instruction, only in `proto` since we _know_ we're going there next.

Reply:
Thanks for explaining, that makes sense. I also have a small question on `cx.nextpc < len(cx.program)`, if it is violating and pointing out of the program, the previous `opB` should throw some error, right?

That is the only question I have on this one.

Reply:
opB allows a jump to the instruction right after the end of the program, as a sort of "jump to done". So we need to check that, since we can't read an instruction there to see if it is a `callsub`

Reply:
I see, if one dumbly make `callsub end`, with `end` pointing to the end of the program, then the execution is essentially ended. To eliminate such cases, we need to add a boundary check for pc, right?

If my think is correct, I am resolving.

Reply:
Almost right. We can _allow_ such a callsub, and in fact we do. But we aren't jumping to a `proto` when that happens, so we don't set the `fromCallsub` flag.